File: benchmarks/BunnymarkV2/es/BunnymarkV2.ts
export default class BunnymarkV2 extends godot.Node2D {
	private grav = 500;
	private bunny_texture = godot.ResourceLoader.load("res://images/godot_bunny.png") as godot.Texture;
	private bunny_speeds = [];
	private label = new godot.Label();
	private bunnies = new godot.Node2D();
	private screen_size: godot.Vector2 = null;

	_ready() {
		this.add_child(this.bunnies);
		this.label.rect_position = new godot.Vector2(0, 20);
		this.add_child(this.label);
	}
	
	_process(delta: number) {
		this.screen_size = this.get_viewport_rect().size;
		this.label.text = "Bunnies: " + this.bunnies.get_child_count();
		let bunny_children: godot.Node2D[] = this.bunnies.get_children();
		for (let i = 0; i < bunny_children.length; i++) {
			const bunny = bunny_children[i];
			let pos = bunny.position;
			let speed = this.bunny_speeds[i];
			pos.x += speed.x * delta;
			pos.y += speed.y * delta;
			speed.y += this.grav * delta;
			
			if (pos.x > this.screen_size.x) {
				speed.x *= -1;
				pos.x = this.screen_size.x;
			}
			
			if (pos.x < 0) {
				speed.x *= -1;
				pos.x = 0;
			}
				
			if (pos.y > this.screen_size.y) {
				pos.y = this.screen_size.y;
				if (Math.random() > 0.5)
					speed.y = -(godot.randi() % 1100 + 50);
				else
					speed.y *= -0.85;
			}
				
			if (pos.y < 0) {
				speed.y = 0
				pos.y = 0
			}
			
			bunny.position = pos;
			this.bunny_speeds[i] = speed;
		}
	}
	
	add_bunny() {
		let bunny = new godot.Sprite();
		bunny.set_texture(this.bunny_texture);
		this.bunnies.add_child(bunny);
		bunny.position = new godot.Vector2(this.screen_size.x / 2, this.screen_size.y / 2);
		this.bunny_speeds.push(new godot.Vector2(godot.randi() % 200 + 50, godot.randi() % 200 + 50));
	}
	
	remove_bunny() {
		let child_count = this.bunnies.get_child_count();
		if (child_count == 0)
			return
		let bunny = this.bunnies.get_child(child_count - 1);
		this.bunny_speeds.pop();
		this.bunnies.remove_child(bunny);
	}

	finish() {
		this.emit_signal("benchmark_finished", this.bunnies.get_child_count());
	}

}

godot.register_class(BunnymarkV2, "BunnymarkV2");

File: benchmarks/bunnymarkv2.rs
use euclid;
use gdnative::*;
use rand;

#[derive(NativeClass)]
#[inherit(Node2D)]
pub struct BunnymarkV2 {
    bunnies: Node2D,
    bunny_speeds: Vec<Vector2>,
    bunny_texture: Texture,
    gravity: f32,
    screen_size: euclid::Size2D<f32>,
    label: Label,
}

#[methods]
impl BunnymarkV2 {
    fn _init(_owner: Node2D) -> Self {
        // NOTE: Direct load call crashes with "handle_crash: Program crashed with signal 11"
        let bunny_texture = unsafe {
            ResourceLoader::godot_singleton()
                /*.load("res://images/godot_bunny.png".into(), "".into(), false)
                .unwrap()
                .cast::<Texture>()
                .unwrap()*/
                .call(
                    "load".into(),
                    &[
                        "res://images/godot_bunny.png".into(),
                        "".into(),
                        false.into(),
                    ],
                )
                .try_to_object::<Texture>()
                .unwrap()
        };
        BunnymarkV2 {
            bunnies: Node2D::new(),
            bunny_speeds: vec![],
            bunny_texture,
            gravity: 500.,
            screen_size: euclid::size2(0., 0.),
            label: Label::new(),
        }
    }

    #[export]
    unsafe fn _ready(&mut self, mut owner: Node2D) {
        owner.set_process(true);
        owner.add_child(Some(self.bunnies.to_object()), false);

        self.label.set_position(Vector2::new(0., 20.));
        owner.add_child(Some(self.label.to_object()), false);
    }

    #[export]
    fn _process(&mut self, owner: Node2D, delta: f32) {
        unsafe {
            self.screen_size = owner.get_viewport_rect().size;
            self.label
                .set_text(format!("Bunnies: {}", self.bunnies.get_child_count()).into());
        }

        let mut bunny_children = unsafe { self.bunnies.get_children() };
        for i in 0..bunny_children.len() as usize {
            if let Some(mut bunny) = bunny_children
                .get_mut_ref(i as i32)
                .try_to_object::<Node2D>()
            {
                let mut position = unsafe { bunny.get_position() };
                let mut speed = self.bunny_speeds[i];

                position.x += speed.x * delta;
                position.y += speed.y * delta;

                speed.y += self.gravity * delta;

                if position.x > self.screen_size.width {
                    speed.x *= -1.;
                    position.x = self.screen_size.width;
                }
                if position.x < 0. {
                    speed.x *= -1.;
                    position.x = 0.;
                }

                if position.y > self.screen_size.height {
                    position.y = self.screen_size.height;
                    if rand::random::<f32>() > 0.5 {
                        speed.y = -((rand::random::<u32>() % 1100 + 50) as f32);
                    } else {
                        speed.y *= -0.85;
                    }
                }

                if position.y < 0. {
                    speed.y = 0.;
                    position.y = 0.;
                }

                unsafe { bunny.set_position(position) };
                self.bunny_speeds[i] = speed;
            }
        }
    }

    #[export]
    unsafe fn add_bunny(&mut self, _owner: Node2D) {
        let mut bunny = Sprite::new();
        bunny.set_texture(Some(self.bunny_texture.clone()));
        self.bunnies.add_child(Some(bunny.to_object()), false);
        bunny.set_position(Vector2::new(
            self.screen_size.width / 2.,
            self.screen_size.height / 2.,
        ));
        self.bunny_speeds.push(Vector2::new(
            (rand::random::<i32>() % 200 + 50) as f32,
            (rand::random::<i32>() % 200 + 50) as f32,
        ));
    }

    #[export]
    unsafe fn remove_bunny(&mut self, _owner: Node2D) {
        let child_count = self.bunnies.get_child_count();
        if child_count == 0 {
            return;
        };
        let bunny = self.bunnies.get_child(child_count - 1);
        if let Some(bunny) = bunny {
            self.bunnies.remove_child(Some(bunny.to_object()));
        }
        self.bunny_speeds.pop();
    }

    #[export]
    unsafe fn finish(&self, mut owner: Node2D) {
        owner.emit_signal(
            "benchmark_finished".into(),
            &[self.bunnies.get_child_count().into()],
        );
    }
}
